docs/cloud/tutorials/gateway-end-user.mdx

```diff
@@ -0,0 +1,184 @@
+
+```shell copy="fl"
+$ fluvio topic create multi-user-demo
+topic "multi-user-demo" created
+```
+
+## Produce records
+
+Produce end user keyed records to the topic using the Fluvio CLI.
+
+```shell copy="cmd"
+$ echo "Message for User1" | fluvio produce multi-user-demo --key "user1"
+$ echo "Message for User2" | fluvio produce multi-user-demo --key "user2"
+```
+
+Confirm that the records were produced successfully with the correct key.
+
+```shell copy="fl"
+$ fluvio consume -Bdk multi-user-demo
+Consuming records from 'multi-user-demo' starting from the beginning of log
+[user1] Message for User1
+[user2] Message for User2
+```
+
+## Set up a Key Filter Server (Rust Example)
+
+Create a simple Key Filter Server that returns a Key Filter based on the End-User Token.
+
+### Setup a new Rust project
+
+Create a HTTP application server using [Actix].
+
+```shell copy="fl"
+$ cargo new key_filter_server
+```
+
+Add the `actix-web` dependency to the `Cargo.toml` file.
+
+```toml
+[package]
+name = "key_filter_server"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+actix-web = "4"
+```
+
+Add the following code to the `src/main.rs` file.
+
+```rust
+use actix_web::{web, App, HttpResponse, HttpServer, Responder};
+
+// Example end user token lookup function
+fn lookup(end_user_token: String) -> Option<String> {
+    if end_user_token == "user1_token_example" {
+        Some("user1".to_string())
+    } else if end_user_token == "user2_token_example" {
+        Some("user2".to_string())
+    } else {
+        None
+    }
+}
+
+// Key Filter endpoint POST handler
+async fn keyfilter(end_user_token: String) -> impl Responder {
+    lookup(end_user_token)
+        .map(|key| HttpResponse::Ok().body(key))
+        .unwrap_or_else(|| HttpResponse::NotFound().finish())
+}
+
+#[actix_web::main]
+async fn main() -> std::io::Result<()> {
+    // Start an HTTP server at 127.0.0.1:4000
+    HttpServer::new(|| {
+        App::new()
+            // Map POST requests on "/keyfilter" to the `keyfilter` function
+            .route("/keyfilter", web::post().to(keyfilter))
+    })
+    .bind(("127.0.0.1", 4000))?
+    .run()
+    .await
+}
+```
+
+### Run the Key Filter Server
+
+Run the Key Filter Server.
+
+```shell copy="fl"
+$ cargo run
+```
+
+Keep this server running in a separate terminal for the duration of this tutorial.
+
+### Expose the Key Filter Server to the Internet using pinggy.io
+
+Expose the Key Filter Server to the internet using pinggy.io. This will allow the WebSocket Gateway to fetch the Key Filter based on the End-User Token.
+
+```shell copy="fl"
+$ ssh -p 443 -R0:localhost:4000 a.pinggy.io
+You are not authenticated.
+Your tunnel will expire in 60 minutes. Upgrade to Pinggy Pro to get unrestricted tunnels. https://dashboard.pinggy.io
+
+http://rnlky-<EXAMPLE>.a.free.pinggy.link
+https://rnlky-<EXAMPLE>.a.free.pinggy.link
+```
+
+Note the assigned public HTTPS URL, which will be used as the Key Filter Remote URL.
```

Comment:
should show how to verify auth callback is working (for exampling using curl).

Reply:
added curl example